### PR TITLE
Jake logs

### PR DESCRIPTION
**Pull request by mavic7**
Got module name working. Is not formatted at the moment, just outputs the literal type name, which imo is good enough. Looking for a code review and permission to merge please

### CODE REVIEW

**Comment by mavic7** [comment]
Also need permission from someone / someone to remove cargo.lock from origin/main and potentially erase history so it does not cause the conflict. cargo.lock has been removed on my end. :)

**Comment by jw2476** [comment]
I had a quick look over and it lgtm, I'll review it properly tomorrow

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
 /target
+.idea
+cargo.lock
```
Comment (jw2476) [inline .gitignore]: I don't know if its case sensitive but should be `Cargo.lock`

File: birb_log/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+pub enum LogCategory {
+    INFO,
+    DEBUG,
+    WARN,
+    ERROR
+}
+pub struct LogEntry {
+    pub module_str: String,
+    pub msg_str: String,
+    pub log_str: String,
```
Comment (jw2476) [inline birb_log/src/lib.rs]: I'd rename these to "module", "message" and "log". Also I'd argue that the formatted string shouldn't be stored in the LogEntry and should be formatted when printed.

File: birb_log/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    pub category: LogCategory
+}
+
+pub type T_Listener = fn(&LogEntry);
```
Comment (jw2476) [inline birb_log/src/lib.rs]: Rename to `Listener` to adhere to Rust code style convention

File: birb_log/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#[derive(Default,Debug)]
+pub struct Log{
+    listeners: Vec<T_Listener>
+}
+
+impl Log {
+    #[must_use]
+    pub fn new() -> Self {
+        let mut log = Self::default();
```
Comment (jw2476) [inline birb_log/src/lib.rs]: Could be a one liner `Self::default()`

File: birb_log/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return log;
+
+    }
+
+   fn translate_mod_uid<T>(&self,object: &T) -> String {
+        return std::any::type_name::<T>().clone().to_string();
```
Comment (jw2476) [inline birb_log/src/lib.rs]: Is the double clone nessesary?

File: birb_log/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    } // temporary - will be changed as necessary later to reflect global module IDs, etc.
+
+   pub fn notify_listeners(&self, log: &LogEntry) {
+        for listener in self.listeners.iter() {
```
Comment (jw2476) [inline birb_log/src/lib.rs]: you can iterate through a list with:
```rs
for listener in &self.listeners {
```

File: birb_log/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            listener(log);
+        }
+    }
+
+    pub fn register_listener(&mut self, func: T_Listener) {
+        self.listeners.push(func);
+    }
+
+    pub fn deregister_listener(&mut self, func: T_Listener) {
+        self.listeners.retain(|x| *x != func);
```
Comment (jw2476) [inline birb_log/src/lib.rs:45]: Has this been tested?

File: birb_log/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    pub fn info<T>(&self, object: &T, msg: String) {
+        let timestamp: u64 = match SystemTime::now().duration_since(SystemTime::UNIX_EPOCH) {
+            Ok(n) => n.as_secs(),
+            Err(_) => panic!("SystemTime before UNIX EPOCH!")
```
Comment (jw2476) [inline birb_log/src/lib.rs]: This could be a call to Result::expect()

File: birb_log/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        };
+        let module_name:String = self.translate_mod_uid(object).clone();
```
Comment (jw2476) [inline birb_log/src/lib.rs]: This is a triple clone from the original type's string

**Review by jw2476 — CHANGES_REQUESTED** [review]
I like the example and the logging API, nice work on getting the type name out.

I've left a couple minor changes. Also can you run this through `cargo fmt` since some of this isn't formatted correctly

**Review by jw2476 — APPROVED** [review]
Fantastic work, we'll circle back to this at some point and make some convenient macros but otherwise lgtm.

**Pull request merged**